src/singletons/Crashpad.cpp

```diff
@@ -0,0 +1,216 @@
+#include <QFile>
+#include <QString>
+
+#ifdef CHATTERINO_WITH_CRASHPAD
+#    include <QApplication>
+
+#    include <memory>
+#    include <string>
+#endif
+
+namespace {
+
+using namespace chatterino;
+using namespace literals;
+
+/// The name of the crashpad handler executable.
+/// This varies across platforms
+#if defined(Q_OS_UNIX)
+const QString CRASHPAD_EXECUTABLE_NAME = QStringLiteral("crashpad_handler");
+#elif defined(Q_OS_WINDOWS)
+const QString CRASHPAD_EXECUTABLE_NAME = QStringLiteral("crashpad_handler.exe");
+#else
+#    error Unsupported platform
+#endif
+
+/// Converts a QString into the platform string representation.
+#if defined(Q_OS_UNIX)
+std::string nativeString(const QString &s)
+{
+    return s.toStdString();
+}
+#elif defined(Q_OS_WINDOWS)
+std::wstring nativeString(const QString &s)
+{
+    return s.toStdWString();
+}
+#else
+#    error Unsupported platform
+#endif
+
+const QString RECOVERY_FILE = u"chatterino-recovery.bin"_s;
+
+/// The recovery options are saved outside the settings
+/// to be able to read them without loading the settings.
+///
+/// The flags are saved in the `RECOVERY_FILE` in their binary representation.
+std::optional<CrashRecovery::Flags> readFlags(const Paths &paths)
+{
+    using Flag = CrashRecovery::Flag;
+
+    QFile file(QDir(paths.crashdumpDirectory).filePath(RECOVERY_FILE));
+    if (!file.open(QFile::ReadOnly))
+    {
+        return std::nullopt;
+    }
+
+    auto line = file.readLine(64);
+
+    static_assert(std::is_same_v<std::underlying_type_t<Flag>, qulonglong>);
+
+    bool ok = false;
+    auto value = static_cast<Flag>(line.toULongLong(&ok));
+    if (!ok)
+    {
+        return std::nullopt;
+    }
+
+    return value;
+}
+
+bool canRestart(const Paths &paths)
+{
+#ifdef NDEBUG
+    const auto &args = chatterino::getArgs();
+    bool noBadArgs =
+        !args.isFramelessEmbed && !args.shouldRunBrowserExtensionHost;
+
+    return noBadArgs && readFlags(paths)
+                            .value_or(CrashRecovery::Flag::None)
+                            .has(CrashRecovery::Flag::DoCrashRecovery);
+#else
+    (void)paths;
+    return false;
+#endif
+}
+
+std::string encodeArguments()
+{
+    std::string args;
+    for (auto arg : getArgs().currentArguments())
+    {
+        if (!args.empty())
+        {
+            args.push_back('+');
+        }
+        args += arg.replace(u'+', u"++"_s).toStdString();
+    }
+    return args;
+}
+
+}  // namespace
+
+namespace chatterino {
+
+using namespace std::string_literals;
+
+void CrashRecovery::initialize(Settings & /*settings*/, Paths &paths)
+{
+    auto flags = readFlags(paths);
+    if (flags)
+    {
+        this->currentFlags_ = *flags;
+    }
+    else
+    {
+        // By default, we don't restart after a crash.
+        this->updateFlags(Flag::None);
+    }
+}
+
+CrashRecovery::Flags CrashRecovery::recoveryFlags() const
+{
+    return this->currentFlags_;
+}
+
+void CrashRecovery::updateFlags(CrashRecovery::Flags flags)
```

Comment:
Let's save & load the crash recovery "settings" as a JSON file instead, just use Qt's JSON functions